app/uk/gov/hmrc/agentclientrelationships/auth/AuthActions.scala

```diff
@@ -80,6 +81,39 @@ trait AuthActions extends AuthorisedFunctions {
   def hasRequiredStrideRole(enrolments: Enrolments, strideRoles: Seq[String]): Boolean =
     strideRoles.exists(s => enrolments.enrolments.exists(_.key == s))
 
+  def determineService(enrolment: Enrolment): Either[Unit, TypeOfEnrolment] =
+    enrolment.key match {
+      case EnrolmentMtdIt.enrolmentKey  => Right(EnrolmentMtdIt)
+      case EnrolmentMtdVat.enrolmentKey => Right(EnrolmentMtdVat)
+      case EnrolmentTrust.enrolmentKey  => Right(EnrolmentTrust)
+      case EnrolmentNino.enrolmentKey   => Right(EnrolmentNino)
+      case e =>
+        Logger(getClass).warn(s"Unsupported service: $e")
+        Left()
+    }
+
+  def withAuthorisedAsClientRequest[A](service: String)(body: Request[AnyContent] => TaxIdentifier => Future[Result])(
+    implicit ec: ExecutionContext): Action[AnyContent] = Action.async { implicit request =>
+    implicit val hc: HeaderCarrier =
+      HeaderCarrierConverter.fromHeadersAndSession(request.headers, Some(request.session))
+
+    authorised(Enrolment(service) and AuthProviders(GovernmentGateway)).retrieve(allEnrolments) { enrolments =>
+      enrolments.getEnrolment(service) match {
+        case Some(e) =>
+          determineService(e) match {
+            case Right(typeOfEnrolment) =>
+              val identifierString = e.getIdentifier(typeOfEnrolment.identifierKey)
+              identifierString.map(i => typeOfEnrolment.identifierForValue(i.value)) match {
+                case Some(id) => body(request)(id)
+                case _        => Future.successful(NoPermissionToPerformOperation)
+              }
+            case Left(_) => Future.successful(NoPermissionToPerformOperation)
+          }
+        case _ => Future.successful(NoPermissionToPerformOperation)
+      }
+    }
+  }
+
   def AuthorisedAsItSaClient[A](implicit ec: ExecutionContext) = AuthorisedAsClient(EnrolmentMtdIt, MtdItId.apply) _
 
   def AuthorisedAsVatClient[A](implicit ec: ExecutionContext) = AuthorisedAsClient(EnrolmentMtdVat, Vrn.apply) _
```

Comment:
so now are we still sing the above two methods ?

Reply:
no